x-pack/legacy/plugins/security/server/lib/authentication/providers/base.ts

```diff
@@ -23,6 +24,7 @@ export interface AuthenticationProviderOptions {
   basePath: string;
   client: Legacy.Plugins.elasticsearch.Cluster;
   log: (tags: string[], message: string) => void;
+  tokens: PublicMethodsOf<Tokens>;
```

Comment:
__note:__ relying on `PublicMethodsOf` makes it super simple to mock `tokens` in TS tests. I hate that tests dictate how we should write code, but here it's more or less logically acceptable.

Reply:
> What you have here seems less intrusive, and makes me wonder if it's a pattern that we should embrace more widely?

In a NP world we won't have too many of these cases (contracts are usually plain objects with public only stuff on them, that's is trivial to mock and work with without additional TS voodoo), hopefully, but in places where we still want something like this I believe we embrace this approach.

Reply:
Yes, the core has the same problems. We pass around ES6 classes and later in tests have type errors passing POJO with the same interface. We haven't decided on the official solution in the core https://github.com/elastic/kibana/issues/33396 
but in some places already started using a separate interface for the contracts https://github.com/skaapgif/kibana/blob/a52d038cbaf605c29a94a3872d4eb70fb60c3fd2/src/legacy/server/saved_objects/service/saved_objects_client.ts#L132
Probably declaring `TokenContract` is more self-describing than `PublicMethodsOf<Tokens>`